operator/pkg/gateway-api/route_checks.go

```diff
@@ -0,0 +1,202 @@
+			}
+			if !allowed {
+				mergeTLSRouteStatusConditions(route, parentRef, []metav1.Condition{
+					tlsRouteAcceptedCondition(route, false, "TLSRoute is not allowed to attach to this Gateway due to namespace selector restrictions"),
+				})
+				return ctrl.Result{}, false, nil
+			}
+		}
+
+		// if gateway allows all namespaces, we do not need to check anything here
+		if *listener.AllowedRoutes.Namespaces.From == gatewayv1beta1.NamespacesFromAll {
+			continue
+		}
+
+		// check if the gateway allows the same namespace as the route
+		if *listener.AllowedRoutes.Namespaces.From == gatewayv1beta1.NamespacesFromSame &&
+			route.GetNamespace() == gw.GetNamespace() {
+			mergeTLSRouteStatusConditions(route, parentRef, []metav1.Condition{
+				tlsRouteAcceptedCondition(route, false, "TLSRoute is not allowed to attach to this Gateway due to namespace restrictions"),
+			})
+			return ctrl.Result{}, false, nil
+		}
+	}
+
+	return ctrl.Result{}, true, nil
+}
+
+func checkGatewayRouteKindAllowed(ctx context.Context, log *logrus.Entry, c client.Client, parentRef gatewayv1alpha2.ParentReference, gw *gatewayv1beta1.Gateway, route *gatewayv1alpha2.TLSRoute) (ctrl.Result, bool, error) {
+	for _, listener := range gw.Spec.Listeners {
+		if listener.AllowedRoutes.Kinds == nil {
+			continue
+		}
+
+		allowed := false
+		routeKind := getGatewayKindForObject(route)
+		for _, kind := range listener.AllowedRoutes.Kinds {
+			if (kind.Group == nil || string(*kind.Group) == gatewayv1beta1.GroupName) &&
+				kind.Kind == kindHTTPRoute && routeKind == kindHTTPRoute {
+				allowed = true
+			} else if (kind.Group == nil || string(*kind.Group) == gatewayv1alpha2.GroupName) &&
+				kind.Kind == kindTLSRoute && routeKind == kindTLSRoute {
+				allowed = true
+			}
+		}
+
+		if !allowed {
+			mergeTLSRouteStatusConditions(route, parentRef, []metav1.Condition{
+				RouteReasonNotAllowedByListeners(route, "TLSRoute is not allowed to attach to this Gateway due to route kind restrictions"),
+			})
+			return ctrl.Result{}, false, nil
+		}
+	}
+
+	return ctrl.Result{}, true, nil
+}
+
+func checkMatchingGatewayHostnames(ctx context.Context, log *logrus.Entry, c client.Client, parentRef gatewayv1alpha2.ParentReference, gw *gatewayv1beta1.Gateway, route *gatewayv1alpha2.TLSRoute) (ctrl.Result, bool, error) {
+	if len(computeHosts(gw, route.Spec.Hostnames)) == 0 {
+		mergeTLSRouteStatusConditions(route, parentRef, []metav1.Condition{
+			tlsNoMatchingListenerHostnameRouteCondition(route, "No matching listener hostname"),
+		})
+
+		return ctrl.Result{}, false, nil
+	}
+
+	return ctrl.Result{}, true, nil
+}
+
+func checkMatchingGatewayPorts(ctx context.Context, log *logrus.Entry, c client.Client, parentRef gatewayv1alpha2.ParentReference, gw *gatewayv1beta1.Gateway, route *gatewayv1alpha2.TLSRoute) (ctrl.Result, bool, error) {
+	if parentRef.Port != nil {
+		found := false
+		for _, listener := range gw.Spec.Listeners {
+			if listener.Port == *parentRef.Port {
+				found = true
```

Comment:
NIT: this and next line can be replaced with `return ctrl.Result{}, true, nil` and then you can get rid of `found` flag.